[#330] Fix new app command order

peaonunes opened the pull request:
### Type: bug fix
Closes: #330

### What are the changes and their implications? :gear:
* Git init happens before any other step in `AppGenerator`.
* Husky hooks will now correctly be applied when commiting files.

### Checklist
- [x] Tests added for changes
- [ ] Any added terminal logging uses `packages/server/src/log.ts`

### Breaking change: no

### Other information

I've took the opportunity to add tests as there were none before. I had to do some tricky mocks though, not super happy.

Locally when I run `yarn test` in the root folder in the project the `cli` tests fail. However, when running inside the `cd packages/cli && yarn test` everything is green. Does anyone know what is the issue?
I had to ignore the test hooks because they were not allowing me to push the code.

peaonunes commented on packages/cli/src/generators/app.ts:
```diff
@@ -81,29 +85,27 @@ class AppGenerator extends Generator<AppGeneratorOptions> {
-      } else {
-        console.error('Failed to run git add')
-      }
+    if (gitInitResult.status === 0) {
```
Curious on the motivation behind not failing if the `git` commands fail. If we want to i am happy to make them throw error here as well as the other commands inside this generator.

aem replied:
I'm not super set on failing if git init fails just because it's not a mandatory part of the setup process. I'll open an issue to discuss this more though, thanks for bringing it up!

peaonunes commented on packages/cli/test/generators/app.test.ts:
```diff
@@ -0,0 +1,237 @@
+        await generator.run()
+      } catch (e) {
+        expect(e).toEqual(new Error('Failed to yarn install'))
+      }
```
Tried hard to make this work with `toThrow`, but seems to be a known issue.
https://github.com/facebook/jest/issues/1700

peaonunes commented on packages/cli/src/generators/app.ts:
```diff
@@ -64,7 +68,7 @@ class AppGenerator extends Generator<AppGeneratorOptions> {
+      throw new Error(`Failed to ${this.options.yarn ? 'yarn' : 'npm'} install`)
     }
 
     console.log(chalk.hex(themeColor).bold('\nDependencies successfully installed.'))
```
Curious why we are using `log` helper and/or `console.log|console.error`. Should it be unified?

aem replied:
I've been incrementally migrating to the `@blitzjs/server/utils/log` helper as I touch relelvant lines. We haven't touched the app generator meaningfully in a bit so I haven't had a chance to migrate these log lines. If you want to migrate these to use the main logging utility it would be a huge help!!

peaonunes replied:
Done https://github.com/blitz-js/blitz/pull/367/commits/c31ba558e752cbeeba225dd0c73b8cb5d01c25b4

aem commented:
Thanks so much for the PR! I hadn't noticed this, great find + fix! 

> Locally when I run yarn test in the root folder in the project the cli tests fail. However, when running inside the cd packages/cli && yarn test everything is green. Does anyone know what is the issue?

This is a known issue, the tests run in parallel in the prepush hook but there's a race condition there :( Since we reuqire a successful build to merge and the CI build runs the tests serially it's ok to ignore them locally, there's no chance it'll make it into the canary branch.

aem commented on packages/cli/src/generators/app.ts:
```diff
@@ -64,7 +68,7 @@ class AppGenerator extends Generator<AppGeneratorOptions> {
     const result = spawn.sync(this.options.yarn ? 'yarn' : 'npm', ['install'], {stdio: 'inherit'})
     if (result.status !== 0) {
-      throw new Error()
+      throw new Error(`Failed to ${this.options.yarn ? 'yarn' : 'npm'} install`)
```
Since we now use the yarn/npm split in a few places, would it be possible to pull that logic out into a constant? It'll prevent any chance of us (most likely me 😬) messing it up in the future

peaonunes replied:
Done 77d4a5c

peaonunes commented on packages/cli/test/generators/app.test.ts:
```diff
@@ -0,0 +1,237 @@
+import * as fs from 'fs-extra'
```
I focused on unit tests here so I tried to mock all the external integrations. I believe we may want to have a good amount of both unit and integration tests.
However, we could add those integrations tests in a different suit file to not conflict setup etc. What do you think of us having some sort of `app.int.test.ts` (I am bad at naming) or even a different folder `cli/test/integration`? 
I am happy to add these as a new PR so we incrementally improve the quality of the tests.

dajinchu replied:
Yeah we used `mock-fs` which worked pretty well.

aem commented on packages/cli/src/generators/app.ts:
```diff
@@ -81,29 +85,27 @@ class AppGenerator extends Generator<AppGeneratorOptions> {
+    for (let command of commands) {
+      const result = spawn.sync(...command)
+      if (result.status !== 0) {
+        console.error(`Fialed to run command ${command[0]} with ${command[1].join(' ')} options.`)
```
```suggestion
        console.error(`Failed to run command ${command[0]} with ${command[1].join(' ')} options.`)
```

peaonunes replied:
Done 77d4a5c0588184c269ec061eb71bbb4d26e61e56

peaonunes commented:
New changes were added while this PR is being reviewed, I will take a time to rebase and fix the conflicts.

flybayer commented:
Thank you @peaonunes! I'll mark as draft — mark it as ready for review once it's updated.

@all-contributors add @peaonunes for code

peaonunes commented:
Hey @flybayer thanks, I've updated now.
@aem this is ready for review again.

Unfortunately seems like a lot have changed from when I started adding tests and refactoring some calls. So I had to undo those. In order to avoid this happening again I've only made the change needed on this PR and added tests related to it. Happy on refactor again portions of this code in the future.

flybayer commented:
@peaonunes did you see your tests are failing?

peaonunes commented:
Hey yeah. Trying to figure out why, its complaining about using wothBrand. My changes have nothing to do with the failing point, but at the same time the code is up to date with canary. I might start fresh from canary again in a few hours.

peaonunes commented:
Hey, everyone. 
Closing this PR , I sorted everything out in this new one https://github.com/blitz-js/blitz/pull/399. 
Sorry for the mess.
cc: @aem @flybayer

peaonunes commented:
Hey, everyone. 
Closing this PR , I sorted everything out in this new one https://github.com/blitz-js/blitz/pull/399. 
Sorry for the mess.
cc: @aem @flybayer
